kart/subprocess_util.py

```diff
@@ -1,12 +1,93 @@
+
+# Every method here - run, call, check_call, check_output, run_and_tee_output, run_then_exit -
+# sets the environment to be the tool_environment() if no other environment is supplied.
+
+# 2. sys.stdin, sys.stdout, sys.stderr: When Kart is run in helper mode, these variables are updated
+# each time a new kart process connects to the helper daemon to be the same as the file-descriptors
+# from the calling process. However, the default stdin,stderr,stdout parameters of subprocess.call
+# are not updated accordingly - that means, when running in helper mode, these parameters need to
+# be set explicitly every time a subprocess is called.
+
+# Every method here - run, call, check_call, check_output, run_and_tee_output, run_then_exit -
+# makes sure to set these parameters explicitly when run in helper mode.
+
+# 3. Capturing output during testing: similar to #2, the click test harness updates sys.stdout
+# and sys.stderr to special values that capture the output so that it can be checked in asserts.
+# Theoretically, the code from #2 would handle this perfectly too, except that unfortunately
+# these special values of sys.stdin and sys.stderr don't have file-descriptors attached and
+# so can't be used in calls to subprocess.run. Since this is only an issue during testing, we
+# instead just set the stdout and stderr to PIPE, capture the subprocess output, and then write
+# it to the special values of sys.stdout and sys.stderr. This breaks real-time progress output
+# for the subprocess but this is not important during testing.
+
+# This fix is only applied to run_and_tee_output and run_then_exit. For correctness, it should be
+# applied to other subprocess calls, but in practice, all of our tests that make asserts about
+# subprocess output rely only on run_then_exit - in the other cases, the test will be unaware
+# of the output from the subprocess, but we don't make any asserts based on it.
+# We could if needed apply this fix to the other methods (with a little more complexity - other
+# methods might have set the stdout / stderr parameters already).
+
+# 4. Two extra functions: run_and_tee_output, run_and_then_exit.
+
+
+CalledProcessError = subprocess.CalledProcessError
+DEVNULL = subprocess.DEVNULL
+PIPE = subprocess.PIPE
+
+
+def run(cmd, **kwargs):
+    return subprocess.run(cmd, **add_default_kwargs(kwargs))
+
+
+def call(cmd, **kwargs):
+    return subprocess.call(cmd, **add_default_kwargs(kwargs))
+
+
+def check_call(cmd, **kwargs):
+    return subprocess.check_call(cmd, **add_default_kwargs(kwargs))
+
+
+def check_output(cmd, **kwargs):
+    return subprocess.check_output(cmd, **add_default_kwargs(kwargs, check_output=True))
+
+
+def Popen(cmd, **kwargs):
+    return subprocess.Popen(cmd, **add_default_kwargs(kwargs))
+
+
+def add_default_kwargs(kwargs_dict, check_output=False):
+    # Set up the environment if not supplied by the caller.
+    if "env" not in kwargs_dict:
+        kwargs_dict.setdefault("env", tool_environment())
```

Comment:
`no_tool_env=True` can currently be requested by explicitly setting `env={}` or `env=os.environ` when calling subprocess_util. However we currently never do this

Reply:
ok. Thoughts on:

> Should we be setting tool_environment all the time, then overriding it with what's in kwargs.env?

Currently it looks like we take `tool_environment()` in a lot of places, sometimes extend it, sometimes seed it from `os.environ`. Is it better to centralise that more?

Reply:
Yeah more could be done here also... at least we hardly ever call tool_environment now - now that the subprocess calls just do it automatically - so that's pretty centralised - but when we do it's cos we need to extend it in some way and the tool_environment isn't as easy to extend as it should be.
I'll add a TODO